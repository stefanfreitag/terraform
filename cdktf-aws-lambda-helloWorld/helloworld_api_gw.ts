import { Construct } from "constructs";
import {
  IamRole,
  LambdaFunction,
  ApiGatewayRestApi,
  ApiGatewayResource,
  ApiGatewayMethod,
  ApiGatewayIntegration,
  LambdaPermission,
  ApiGatewayDeployment,
  DataAwsRegion,
  DataAwsCallerIdentity,
} from "./.gen/providers/aws";

export class HelloWorldApiGw extends Construct {
  constructor(scope: Construct, name: string) {
    super(scope, name);

    const role = new IamRole(this, "basic_lambda_role", {
      description: "Basic Lambda Execution Role",
      assumeRolePolicy: `{ "Version": "2012-10-17", "Statement": [ {
       "Action": "sts:AssumeRole", "Principal": { "Service": "lambda.amazonaws.com" }, "Effect": "Allow", "Sid": "" } ] }`,
    });
    const fn = new LambdaFunction(this, "lambda-hello-world", {
      functionName: "helloWorld",
      handler: "hello_world.lambda_handler",
      runtime: "python3.7",
      description: "A simple Hello world Lambda Function",
      role: role.arn,
      timeout: 60,
      filename: "../hello_world.zip",
    });
    const api = new ApiGatewayRestApi(this, "api-gateway", {
      name: "rest-api",
    });
    const resource = new ApiGatewayResource(this, "api-gateway-resource", {
      restApiId: api.id!,
      parentId: api.rootResourceId,
      pathPart: "hello",
    });
    const method = new ApiGatewayMethod(this, "api-gateway-method", {
      restApiId: api.id!,
      resourceId: resource.id!,
      httpMethod: "GET",
      authorization: "NONE",
    });
    new ApiGatewayIntegration(this, "api-gateway-integration", {
      restApiId: api.id!,
      resourceId: resource.id!,
      httpMethod: method.httpMethod,
      integrationHttpMethod: "POST",
      type: "AWS_PROXY",
      uri: fn.invokeArn,
    });

    const region = new DataAwsRegion(this, "region");
    const userId = new DataAwsCallerIdentity(this, "userId");

    new LambdaPermission(this, "apigw-lambda-permission", {
      action: "lambda:InvokeFunction",
      principal: "apigateway.amazonaws.com",
      functionName: fn.functionName,
      sourceArn:
        "arn:aws:execute-api:" +
        region.name +
        ":" +
        userId.accountId +
        ":" +
        api.id +
        "/*/" +
        method.httpMethod +
        resource.path,
    });

    new ApiGatewayDeployment(this, "api-gateway-deployment", {
      restApiId: api.id!,
      stageName: "Production",
      stageDescription: "Production Environment",
      description: "Hello World - Production environment deployment",
    });

    /**
    new TerraformOutput(this, "endpoint", {
      value: deployment.invokeUrl
    });
     */
  }
}
