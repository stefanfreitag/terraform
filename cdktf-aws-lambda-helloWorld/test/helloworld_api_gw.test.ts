
import { Testing, TerraformStack } from "cdktf";
import { HelloWorldApiGw} from "../helloworld_api_gw";



test("snapshot test", () => {  
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");
  new HelloWorldApiGw(stack, "api-gw");
  
  expect(Testing.synth(stack)).toMatchSnapshot();
});