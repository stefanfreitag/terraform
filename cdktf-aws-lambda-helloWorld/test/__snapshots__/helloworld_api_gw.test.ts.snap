// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_iam_role\\": {
      \\"test_apigw_basiclambdarole_B5B06285\\": {
        \\"assume_role_policy\\": \\"{ \\\\\\"Version\\\\\\": \\\\\\"2012-10-17\\\\\\", \\\\\\"Statement\\\\\\": [ {\\\\n       \\\\\\"Action\\\\\\": \\\\\\"sts:AssumeRole\\\\\\", \\\\\\"Principal\\\\\\": { \\\\\\"Service\\\\\\": \\\\\\"lambda.amazonaws.com\\\\\\" }, \\\\\\"Effect\\\\\\": \\\\\\"Allow\\\\\\", \\\\\\"Sid\\\\\\": \\\\\\"\\\\\\" } ] }\\",
        \\"description\\": \\"Basic Lambda Execution Role\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/basic_lambda_role\\",
            \\"uniqueId\\": \\"test_apigw_basiclambdarole_B5B06285\\"
          }
        }
      }
    },
    \\"aws_lambda_function\\": {
      \\"test_apigw_lambdahelloworld_E21B6A04\\": {
        \\"description\\": \\"A simple Hello world Lambda Function\\",
        \\"filename\\": \\"../hello_world.zip\\",
        \\"function_name\\": \\"helloWorld\\",
        \\"handler\\": \\"hello_world.lambda_handler\\",
        \\"role\\": \\"\${aws_iam_role.test_apigw_basiclambdarole_B5B06285.arn}\\",
        \\"runtime\\": \\"python3.7\\",
        \\"timeout\\": 60,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/lambda-hello-world\\",
            \\"uniqueId\\": \\"test_apigw_lambdahelloworld_E21B6A04\\"
          }
        }
      }
    },
    \\"aws_api_gateway_rest_api\\": {
      \\"test_apigw_apigateway_4E70AAF9\\": {
        \\"name\\": \\"rest-api\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/api-gateway\\",
            \\"uniqueId\\": \\"test_apigw_apigateway_4E70AAF9\\"
          }
        }
      }
    },
    \\"aws_api_gateway_resource\\": {
      \\"test_apigw_apigatewayresource_D056EDEA\\": {
        \\"parent_id\\": \\"\${aws_api_gateway_rest_api.test_apigw_apigateway_4E70AAF9.root_resource_id}\\",
        \\"path_part\\": \\"hello\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.test_apigw_apigateway_4E70AAF9.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/api-gateway-resource\\",
            \\"uniqueId\\": \\"test_apigw_apigatewayresource_D056EDEA\\"
          }
        }
      }
    },
    \\"aws_api_gateway_method\\": {
      \\"test_apigw_apigatewaymethod_D543AC51\\": {
        \\"authorization\\": \\"NONE\\",
        \\"http_method\\": \\"GET\\",
        \\"resource_id\\": \\"\${aws_api_gateway_resource.test_apigw_apigatewayresource_D056EDEA.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.test_apigw_apigateway_4E70AAF9.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/api-gateway-method\\",
            \\"uniqueId\\": \\"test_apigw_apigatewaymethod_D543AC51\\"
          }
        }
      }
    },
    \\"aws_api_gateway_integration\\": {
      \\"test_apigw_apigatewayintegration_F4EDBBD6\\": {
        \\"http_method\\": \\"GET\\",
        \\"integration_http_method\\": \\"POST\\",
        \\"resource_id\\": \\"\${aws_api_gateway_resource.test_apigw_apigatewayresource_D056EDEA.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.test_apigw_apigateway_4E70AAF9.id}\\",
        \\"type\\": \\"AWS_PROXY\\",
        \\"uri\\": \\"\${aws_lambda_function.test_apigw_lambdahelloworld_E21B6A04.invoke_arn}\\",
        \\"depends_on\\": [
          \\"aws_api_gateway_method.test_apigw_apigatewaymethod_D543AC51\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/api-gateway-integration\\",
            \\"uniqueId\\": \\"test_apigw_apigatewayintegration_F4EDBBD6\\"
          }
        }
      }
    },
    \\"aws_lambda_permission\\": {
      \\"test_apigw_apigwlambdapermission_6E2E14F7\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"function_name\\": \\"helloWorld\\",
        \\"principal\\": \\"apigateway.amazonaws.com\\",
        \\"source_arn\\": \\"arn:aws:execute-api:\${data.aws_region.test_apigw_region_75D4FC1C.name}:\${data.aws_caller_identity.test_apigw_userId_A866FA9A.account_id}:\${aws_api_gateway_rest_api.test_apigw_apigateway_4E70AAF9.id}/*/GET\${aws_api_gateway_resource.test_apigw_apigatewayresource_D056EDEA.path}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/apigw-lambda-permission\\",
            \\"uniqueId\\": \\"test_apigw_apigwlambdapermission_6E2E14F7\\"
          }
        }
      }
    },
    \\"aws_api_gateway_deployment\\": {
      \\"test_apigw_apigatewaydeployment_8BAE2D89\\": {
        \\"description\\": \\"Hello World - Production environment deployment\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.test_apigw_apigateway_4E70AAF9.id}\\",
        \\"stage_description\\": \\"Production Environment\\",
        \\"stage_name\\": \\"Production\\",
        \\"depends_on\\": [
          \\"aws_api_gateway_method.test_apigw_apigatewaymethod_D543AC51\\",
          \\"aws_api_gateway_integration.test_apigw_apigatewayintegration_F4EDBBD6\\"
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/api-gateway-deployment\\",
            \\"uniqueId\\": \\"test_apigw_apigatewaydeployment_8BAE2D89\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"aws_region\\": {
      \\"test_apigw_region_75D4FC1C\\": {
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/region\\",
            \\"uniqueId\\": \\"test_apigw_region_75D4FC1C\\"
          }
        }
      }
    },
    \\"aws_caller_identity\\": {
      \\"test_apigw_userId_A866FA9A\\": {
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/api-gw/userId\\",
            \\"uniqueId\\": \\"test_apigw_userId_A866FA9A\\"
          }
        }
      }
    }
  },
  \\"output\\": {
    \\"test_apigw_endpoint_5C407DAB\\": {
      \\"value\\": \\"\${aws_api_gateway_deployment.test_apigw_apigatewaydeployment_8BAE2D89.invoke_url}\\"
    }
  }
}"
`;
